File: python/evm_state/proof.py
"""Account and complete-storage verification against an explicitly chosen state root.

This verifies trie proofs, not BSC consensus. The caller records how it trusts
the finalized block header; a provider-supplied header is a distinct trust source.
"""
from dataclasses import dataclass
import re

from Crypto.Hash import keccak
import rlp
from rlp.codec import encode_raw
from trie import HexaryTrie
from trie.utils.nibbles import bytes_to_nibbles, decode_nibbles, encode_nibbles

from .triedb import StorageSortDB

EMPTY_STORAGE_ROOT = bytes.fromhex("56e81f171bcc55a6ff8345e692c0f86e5b48e01b996cadc001622fb5e363b421")


class VerificationError(ValueError):
    pass


def keccak256(value: bytes) -> bytes:
    return keccak.new(digest_bits=256, data=value).digest()


EMPTY_CODE_HASH = keccak256(b"")


def unhex(value: str, size: int | None = None) -> bytes:
    if not isinstance(value, str) or not re.fullmatch(r"0x(?:[0-9a-fA-F]{2})*", value):
        raise VerificationError("expected an even-length 0x-prefixed hex string")
    decoded = bytes.fromhex(value[2:])
    if size is not None and len(decoded) != size:
        raise VerificationError(f"expected {size} bytes, got {len(decoded)}")
    return decoded


def address(value: str) -> str:
    return "0x" + unhex(value, 20).hex()


def quantity(value: str, bits: int = 256) -> int:
    if not isinstance(value, str) or not re.fullmatch(r"0x[0-9a-fA-F]+", value):
        raise VerificationError("expected a JSON-RPC hex quantity")
    parsed = int(value, 16)
    if parsed >= 2**bits:
        raise VerificationError(f"quantity exceeds uint{bits}")
    return parsed


@dataclass(frozen=True)
class Account:
    address: str
    exists: bool
    nonce: int
    balance: int
    storage_root: bytes
    code_hash: bytes

    def json(self) -> dict:
        return {
            "address": self.address, "exists": self.exists, "nonce": self.nonce,
            "balance": str(self.balance), "storage_root": "0x" + self.storage_root.hex(),
            "code_hash": "0x" + self.code_hash.hex(),
        }


def verify_account(state_root: str, expected_address: str, proof: dict) -> Account:
    """Verify inclusion/non-inclusion and every RPC-reported account field."""
    root = unhex(state_root, 32)
    expected_address = address(expected_address)
    if address(proof.get("address", expected_address)) != expected_address:
        raise VerificationError("proof is for a different account")
    try:
        nodes = tuple(rlp.decode(unhex(node)) for node in proof["accountProof"])
        encoded = HexaryTrie.get_from_proof(root, keccak256(unhex(expected_address, 20)), nodes)
    except Exception as error:
        # py-trie raises several validation/missing-node classes. Keep one public
        # error boundary; do not downgrade invalid or incomplete proofs to absence.
        raise VerificationError("invalid or incomplete account proof") from error
    if encoded:
        try:
            fields = rlp.decode(encoded)
            if not isinstance(fields, list) or len(fields) != 4:
                raise ValueError("account must have four fields")
            nonce_bytes, balance_bytes, storage_root, code_hash = fields
            if any(not isinstance(item, bytes) for item in fields):
                raise ValueError("invalid account field encoding")
            if nonce_bytes.startswith(b"\0") or balance_bytes.startswith(b"\0"):
                raise ValueError("non-canonical account integer")
            nonce, balance = int.from_bytes(nonce_bytes, "big"), int.from_bytes(balance_bytes, "big")
            if nonce >= 2**64 or balance >= 2**256 or len(storage_root) != 32 or len(code_hash) != 32:
                raise ValueError("invalid account field length")
            account = Account(expected_address, True, nonce, balance, storage_root, code_hash)
        except Exception as error:
            raise VerificationError("invalid account trie value") from error
    else:
        account = Account(expected_address, False, 0, 0, EMPTY_STORAGE_ROOT, EMPTY_CODE_HASH)
    try:
        reported = (quantity(proof["nonce"], 64), quantity(proof["balance"]),
                    unhex(proof["storageHash"], 32), unhex(proof["codeHash"], 32))
    except (KeyError, TypeError) as error:
        raise VerificationError("missing account proof metadata") from error
    if reported != (account.nonce, account.balance, account.storage_root, account.code_hash):
        raise VerificationError("RPC account metadata does not match the proven account")
    return account


def _ordered_storage_root(entries) -> bytes:
    """Build the MPT root of ordered, unique 32-byte hashed storage keys.

    Values are already RLP-encoded nonzero storage integers. Two lookahead
    entries identify leaves; completed siblings collapse to inline nodes or
    hashes. The fixed key length bounds recursion at 64 nibbles and means a
    branch never has a terminal value. No full trie or node reference counts
    remain in memory. Compact paths and RLP use the pinned py-trie/rlp codecs.
    Encoding rules: https://ethereum.org/en/developers/docs/data-structures-and-encoding/patricia-merkle-trie/
    """
    source, previous = iter(entries), None

    def take():
        nonlocal previous
        try:
            key, value = next(source)
        except StopIteration:
            return None
        if (not isinstance(key, bytes) or len(key) != 32
                or (previous is not None and key <= previous)):
            raise VerificationError("expected unique ordered 32-byte hashed storage keys")
        previous = key
        return bytes_to_nibbles(key), value

    current, following = take(), take()
    if current is None:
        return EMPTY_STORAGE_ROOT

    def advance():
        nonlocal current, following
        current, following = following, take()

    def reference(node):
        encoded = encode_raw(node)
        return node if len(encoded) < 32 else keccak256(encoded)

    def subtree(depth):
        prefix = current[0][:depth]
        if following is None or following[0][:depth] != prefix:
            key, value = current
            advance()
            return [encode_nibbles(key[depth:] + (16,)), value]
        children = []
        while current is not None and current[0][:depth] == prefix:
            digit = current[0][depth]
            children.append((digit, subtree(depth + 1)))
        if len(children) == 1:
            digit, child = children[0]
            if len(child) == 2:
                return [encode_nibbles((digit,) + decode_nibbles(child[0])), child[1]]
            return [encode_nibbles((digit,)), reference(child)]
        branch = [b""] * 17
        for digit, child in children:
            branch[digit] = reference(child)
        return branch

    return keccak256(encode_raw(subtree(0)))


def storage_root(slots, database=None) -> tuple[bytes, int]:
    """Hash all nonzero (32-byte slot, 32-byte value) pairs.

    Supply a fresh StorageSortDB for large reconstructions. The original
    HexaryTrie path remains available with a byte mapping or no database.
    Both paths reject duplicates even when their values agree.
    """
    if isinstance(database, StorageSortDB):
        database.begin()
        count = 0
        for slot, value in slots:
            key = keccak256(unhex(slot, 32))
            number = int.from_bytes(unhex(value, 32), "big")
            if number == 0:
                raise VerificationError("complete-storage input must contain nonzero slots only")
            try:
                database.add(key, rlp.encode(number))
            except KeyError as error:
                raise VerificationError("duplicate nonzero storage slot") from error
            count += 1
        return _ordered_storage_root(database.ordered_entries()), count
    tree = HexaryTrie({} if database is None else database, prune=True)
    count = 0
    for slot, value in slots:
        slot_bytes, value_bytes = unhex(slot, 32), unhex(value, 32)
        key = keccak256(slot_bytes)
        if tree.get(key):
            raise VerificationError("duplicate nonzero storage slot")
        number = int.from_bytes(value_bytes, "big")
        if number == 0:
            raise VerificationError("complete-storage input must contain nonzero slots only")
        tree[key] = rlp.encode(number)
        count += 1
    return tree.root_hash, count


def verify_metadata(account: Account, code: str, metadata: dict):
    """Require every account field and actual bytecode to match its proven leaf."""
    try:
        nonce = metadata["nonce"]
        balance = metadata["balance"]
        if isinstance(nonce, bool) or not isinstance(nonce, int):
            raise VerificationError("nonce must be an exact integer")
        if not isinstance(balance, str) or not re.fullmatch(r"0|[1-9][0-9]*", balance):
            raise VerificationError("balance must be an exact decimal string")
        if nonce != account.nonce or int(balance) != account.balance:
            raise VerificationError("account nonce/balance mismatch")
        if unhex(metadata["code_hash"], 32) != account.code_hash:
            raise VerificationError("account code hash mismatch")
    except (KeyError, TypeError) as error:
        raise VerificationError("missing required account metadata") from error
    if keccak256(unhex(code)) != account.code_hash:
        raise VerificationError("bytecode does not match the proven code hash")


def verify_complete(account: Account, slots, code: str, metadata: dict, database=None) -> int:
    """Fail on a missing field, wrong account state, incomplete storage or wrong code."""
    verify_metadata(account, code, metadata)
    root, count = storage_root(slots, database)
    if root != account.storage_root:
        raise VerificationError("storage root mismatch: incomplete, stale or incorrect account storage")
    return count

File: crates/evm-state/src/proof.rs
//! Proofs bind complete account state to a chosen root, not to BSC consensus.
use alloy_primitives::{keccak256, Bytes, B256, U256};
use alloy_trie::{proof::verify_proof, HashBuilder, Nibbles, EMPTY_ROOT_HASH};
use anyhow::{bail, ensure, Context, Result};
use rusqlite::{params, Connection};
use serde_json::{json, Value};
use std::path::Path;

pub fn unhex(value: &str) -> Result<Vec<u8>> {
    let raw = value
        .strip_prefix("0x")
        .context("expected 0x-prefixed hex")?;
    ensure!(raw.len() % 2 == 0, "expected even-length hex");
    Ok(hex::decode(raw).context("invalid hex")?)
}

pub fn fixed<const N: usize>(value: &str) -> Result<[u8; N]> {
    unhex(value)?
        .try_into()
        .map_err(|_| anyhow::anyhow!("expected {N} bytes"))
}

pub fn string<'a>(value: &'a Value, field: &str) -> Result<&'a str> {
    value
        .get(field)
        .and_then(Value::as_str)
        .with_context(|| format!("missing string field {field}"))
}

pub fn address(value: &str) -> Result<String> {
    Ok(format!("0x{}", hex::encode(fixed::<20>(value)?)))
}

pub fn quantity(value: &str, bits: usize) -> Result<U256> {
    let raw = value
        .strip_prefix("0x")
        .context("expected JSON-RPC hex quantity")?;
    ensure!(
        !raw.is_empty() && raw.bytes().all(|v| v.is_ascii_hexdigit()),
        "invalid JSON-RPC quantity"
    );
    let number = U256::from_str_radix(raw, 16).context("quantity exceeds uint256")?;
    ensure!(number.bit_len() <= bits, "quantity exceeds uint{bits}");
    Ok(number)
}

pub fn rlp_list(encoded_items: &[Vec<u8>]) -> Vec<u8> {
    let mut out = Vec::new();
    alloy_rlp::Header {
        list: true,
        payload_length: encoded_items.iter().map(Vec::len).sum(),
    }
    .encode(&mut out);
    for item in encoded_items {
        out.extend_from_slice(item);
    }
    out
}

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct Account {
    pub address: String,
    pub exists: bool,
    pub nonce: u64,
    pub balance: U256,
    pub storage_root: B256,
    pub code_hash: B256,
}

impl Account {
    pub fn json(&self) -> Value {
        json!({"address": self.address, "exists": self.exists, "nonce": self.nonce,
            "balance": self.balance.to_string(), "storage_root": format!("{:#x}", self.storage_root),
            "code_hash": format!("{:#x}", self.code_hash)})
    }
}

pub fn verify_account(state_root: &str, selected: &str, proof: &Value) -> Result<Account> {
    let root = B256::from(fixed::<32>(state_root)?);
    let selected = address(selected)?;
    if let Some(actual) = proof.get("address") {
        ensure!(
            address(actual.as_str().context("invalid proof address")?)? == selected,
            "proof is for a different account"
        );
    }
    let nonce: u64 = quantity(string(proof, "nonce")?, 64)?
        .try_into()
        .context("nonce exceeds uint64")?;
    let balance = quantity(string(proof, "balance")?, 256)?;
    let storage_root = B256::from(fixed::<32>(string(proof, "storageHash")?)?);
    let code_hash = B256::from(fixed::<32>(string(proof, "codeHash")?)?);
    let nodes = proof
        .get("accountProof")
        .and_then(Value::as_array)
        .context("missing account proof")?
        .iter()
        .map(|node| {
            Ok(Bytes::from(unhex(
                node.as_str().context("invalid account proof node")?,
            )?))
        })
        .collect::<Result<Vec<_>>>()?;
    let key = keccak256(fixed::<20>(&selected)?);
    // Verifying the canonical encoding of all reported fields against the root
    // gives them no authority of their own. A different field cannot pass.
    let encoded = rlp_list(&[
        alloy_rlp::encode(nonce),
        alloy_rlp::encode(balance),
        alloy_rlp::encode(storage_root.as_slice()),
        alloy_rlp::encode(code_hash.as_slice()),
    ]);
    let exists = if verify_proof(root, Nibbles::unpack(key), Some(encoded), &nodes).is_ok() {
        true
    } else {
        verify_proof(root, Nibbles::unpack(key), None, &nodes)
            .context("invalid or incomplete account proof")?;
        ensure!(
            nonce == 0
                && balance.is_zero()
                && storage_root == EMPTY_ROOT_HASH
                && code_hash == keccak256([]),
            "RPC account metadata does not match proven absence"
        );
        false
    };
    Ok(Account {
        address: selected,
        exists,
        nonce,
        balance,
        storage_root,
        code_hash,
    })
}

pub fn verify_metadata(account: &Account, code: &str, metadata: &Value) -> Result<()> {
    let nonce = metadata
        .get("nonce")
        .and_then(Value::as_u64)
        .context("nonce must be an exact uint64")?;
    let balance = string(metadata, "balance")?;
    ensure!(
        balance == "0"
            || (!balance.starts_with('0')
                && !balance.is_empty()
                && balance.bytes().all(|v| v.is_ascii_digit())),
        "balance must be an exact decimal string"
    );
    ensure!(
        nonce == account.nonce && U256::from_str_radix(balance, 10)? == account.balance,
        "account nonce/balance mismatch"
    );
    ensure!(
        B256::from(fixed::<32>(string(metadata, "code_hash")?)?) == account.code_hash,
        "account code hash mismatch"
    );
    ensure!(
        keccak256(unhex(code)?) == account.code_hash,
        "bytecode does not match proven code hash"
    );
    Ok(())
}

/// A single-use disk sort. No full account map or full trie is kept in memory.
pub struct StorageSort {
    connection: Connection,
    count: u64,
    failed: bool,
}

impl StorageSort {
    pub fn new(path: &Path) -> Result<Self> {
        let connection = Connection::open(path)?;
        connection.execute_batch(
            "PRAGMA journal_mode=OFF; PRAGMA synchronous=OFF; PRAGMA cache_size=-16384;
            CREATE TABLE storage (key BLOB PRIMARY KEY, value BLOB NOT NULL) WITHOUT ROWID; BEGIN",
        )?;
        Ok(Self {
            connection,
            count: 0,
            failed: false,
        })
    }

    pub fn insert(&mut self, slot: &str, value: &str) -> Result<()> {
        ensure!(!self.failed, "storage sorting workspace has failed");
        self.failed = true;
        let key = keccak256(fixed::<32>(slot)?);
        let value = U256::from_be_bytes(fixed::<32>(value)?);
        ensure!(
            !value.is_zero(),
            "complete-storage input must contain nonzero slots only"
        );
        let encoded = alloy_rlp::encode(value);
        let inserted = self
            .connection
            .prepare_cached("INSERT INTO storage VALUES (?1,?2)")?
            .execute(params![key.as_slice(), encoded]);
        if let Err(error) = inserted {
            if matches!(&error, rusqlite::Error::SqliteFailure(v, _) if v.code == rusqlite::ErrorCode::ConstraintViolation)
            {
                bail!("duplicate nonzero storage slot");
            }
            return Err(error.into());
        }
        self.count += 1;
        self.failed = false;
        Ok(())
    }

    pub fn finish(self) -> Result<(B256, u64)> {
        ensure!(!self.failed, "storage sorting workspace has failed");
        self.connection.execute_batch("COMMIT")?;
        let mut statement = self
            .connection
            .prepare("SELECT key,value FROM storage ORDER BY key")?;
        let mut rows = statement.query([])?;
        let mut builder = HashBuilder::default();
        let mut previous: Option<[u8; 32]> = None;
        let mut observed = 0_u64;
        while let Some(row) = rows.next()? {
            let key: Vec<u8> = row.get(0)?;
            let key: [u8; 32] = key
                .try_into()
                .map_err(|_| anyhow::anyhow!("invalid staged storage key"))?;
            ensure!(
                previous.is_none_or(|last| key > last),
                "unordered or duplicate staged storage key"
            );
            previous = Some(key);
            let value: Vec<u8> = row.get(1)?;
            builder.add_leaf(Nibbles::unpack(key), &value);
            observed += 1;
        }
        ensure!(observed == self.count, "staged storage count changed");
        Ok((builder.root(), observed))
    }
}

pub fn storage_root<I>(slots: I, path: &Path) -> Result<(B256, u64)>
where
    I: IntoIterator<Item = Result<(String, String)>>,
{
    let mut database = StorageSort::new(path)?;
    for pair in slots {
        let (slot, value) = pair?;
        database.insert(&slot, &value)?;
    }
    database.finish()
}

pub fn verify_complete<I>(
    account: &Account,
    slots: I,
    code: &str,
    metadata: &Value,
    path: &Path,
) -> Result<u64>
where
    I: IntoIterator<Item = Result<(String, String)>>,
{
    verify_metadata(account, code, metadata)?;
    let (root, count) = storage_root(slots, path)?;
    ensure!(
        root == account.storage_root,
        "storage root mismatch: incomplete, stale or incorrect account storage"
    );
    Ok(count)
}
